internal/core/src/index/BitmapIndex.cpp

```diff
@@ -0,0 +1,648 @@
+    for (auto& pair : data_) {
+        memcpy(data_ptr, &pair.first, sizeof(T));
+        data_ptr += sizeof(T);
+
+        pair.second.write(reinterpret_cast<char*>(data_ptr));
+        data_ptr += pair.second.getSizeInBytes();
+    }
+}
+
+template <>
+void
+BitmapIndex<std::string>::SerializeIndexData(uint8_t* data_ptr) {
+    for (auto& pair : data_) {
+        size_t key_size = pair.first.size();
+        memcpy(data_ptr, &key_size, sizeof(size_t));
+        data_ptr += sizeof(size_t);
+
+        memcpy(data_ptr, pair.first.data(), key_size);
+        data_ptr += key_size;
+
+        pair.second.write(reinterpret_cast<char*>(data_ptr));
+        data_ptr += pair.second.getSizeInBytes();
+    }
+}
+
+template <typename T>
+BinarySet
+BitmapIndex<T>::Serialize(const Config& config) {
+    AssertInfo(is_built_, "index has not been built yet");
+
+    auto index_data_size = 0;
+    for (auto& pair : data_) {
+        index_data_size += pair.second.getSizeInBytes() + sizeof(T);
+    }
+
+    std::shared_ptr<uint8_t[]> index_data(new uint8_t[index_data_size]);
+    uint8_t* data_ptr = index_data.get();
+    SerializeIndexData(data_ptr);
+
+    std::shared_ptr<uint8_t[]> index_length(new uint8_t[sizeof(size_t)]);
+    auto index_size = data_.size();
+    memcpy(index_length.get(), &index_size, sizeof(size_t));
+
+    std::shared_ptr<uint8_t[]> num_rows(new uint8_t[sizeof(size_t)]);
+    memcpy(num_rows.get(), &total_num_rows_, sizeof(size_t));
+
+    BinarySet ret_set;
+    ret_set.Append(BITMAP_INDEX_DATA, index_data, index_data_size);
+    ret_set.Append(BITMAP_INDEX_LENGTH, index_length, sizeof(size_t));
+    ret_set.Append(BITMAP_INDEX_NUM_ROWS, num_rows, sizeof(size_t));
+
+    LOG_INFO("build bitmap index with cardinality = {}, num_rows = {}",
+             index_size,
+             total_num_rows_);
+    return ret_set;
+}
+
+template <typename T>
+BinarySet
+BitmapIndex<T>::Upload(const Config& config) {
+    auto binary_set = Serialize(config);
+
+    file_manager_->AddFile(binary_set);
+
+    auto remote_path_to_size = file_manager_->GetRemotePathsToFileSize();
+    BinarySet ret;
+    for (auto& file : remote_path_to_size) {
+        ret.Append(file.first, nullptr, file.second);
+    }
+    return ret;
+}
+
+template <typename T>
+BinarySet
+BitmapIndex<T>::UploadV2(const Config& config) {
+    auto binary_set = Serialize(config);
+
+    file_manager_->AddFileV2(binary_set);
+
+    auto remote_path_to_size = file_manager_->GetRemotePathsToFileSize();
+    BinarySet ret;
+    for (auto& file : remote_path_to_size) {
+        ret.Append(file.first, nullptr, file.second);
+    }
+    return ret;
+}
+
+template <typename T>
+void
+BitmapIndex<T>::Load(const BinarySet& binary_set, const Config& config) {
+    milvus::Assemble(const_cast<BinarySet&>(binary_set));
+    LoadWithoutAssemble(binary_set, config);
+}
+
+template <typename T>
+TargetBitmap
+BitmapIndex<T>::ConvertRoaringToBitset(const roaring::Roaring& values) {
```

Comment:
Consider to pass the reference or pointer of TargetBitmap to avoid copy.

Reply:
compiler can optimize it using move